### PR TITLE
698: Whitelabel PDF prints

### PR DESCRIPTION
**Pull request by maxammann**
related to #698

### CODE REVIEW

**Comment by maxammann** [comment]
Ready for review, templates for Nürnberg and Bavaria will follow.

File: administration/src/cards/PdfFactory.ts [hunk 1]
```diff
@@ -1,167 +1,150 @@
...
-    [
-      'Anleitung:',
-      '1. Laden Sie sich die App "Ehrenamtskarte" herunter.',
-      '2. Starten Sie die App und folgen Sie den Hinweisen zum Scannen des Anmeldecodes.',
-      '3. Scannen Sie den Anmeldecode.',
-    ],
-    pageMargin,
-    instructionsY,
-    { baseline: 'middle' }
+  // Static QR code
```
Comment (michael-markl) [inline administration/src/cards/PdfFactory.ts:53]:
```suggestion
  // TODO: Plug in the Static QR code here. Don't print static qr code for bayern
```

File: administration/src/util/qrcode.ts [hunk 2]
```diff
@@ -84,13 +84,23 @@ export const drawjsPDF = (
...
-        pdfDocument.setLineWidth(0.4)
-        pdfDocument.roundedRect(x + rectX, y + rectY, rectWidth, rectHeight, 10, 10, 'D')
-        pdfDocument.setLineWidth(0)
+        pdfDocument.drawRectangle({
+          x: x + rectX,
+          y: y + rectY,
+          width: rectWidth,
+          height: rectHeight,
+          borderWidth: 1,
+          color: rgb(1, 1, 1),
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:102]: Do we really want a border?

**Review by michael-markl — APPROVED** [review]

File: administration/src/cards/PdfFactory.ts [hunk 3]
```diff
@@ -1,167 +1,150 @@
...
-    const output = doc.output('blob')
-    return new Blob([output], { type: 'application/pdf' })
-  } catch {
-    throw new Exception({
-      type: 'pdf-generation',
-    })
+    await fillContentAreas(doc, page, activationCode, region)
   }
+
+  doc.setTitle('Karten')
```
[the remark this answers is not in the document]
Comment (maxammann) [reply administration/src/cards/PdfFactory.ts:145]: It's nice if the doc has at least some title. It is displayed in the browser tabs for instance.

File: administration/src/cards/PdfFactory.ts [hunk 4]
```diff
@@ -1,167 +1,150 @@
...
+function fillCodeArea(
+  activationCode:
+    | {
+        value: DynamicActivationCode
+        case: 'dynamicActivationCode'
+      }
+    | {
+        value: StaticVerifyCode
+        case: 'staticVerifyCode'
+      },
```
[the remark this answers is not in the document]
Comment (maxammann) [reply administration/src/cards/PdfFactory.ts:109]: Only static verify codes and activation codes are allowed. So this is safer.

File: administration/src/cards/PdfFactory.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (michael-markl) [reply administration/src/cards/PdfFactory.ts:146]: Then, we should probably whitelabel it.

File: administration/src/cards/PdfFactory.ts [hunk 5]
```diff
@@ -95,49 +92,27 @@ Aussteller: ${region.prefix} ${region.name}`,
...
     }).toBinary()
   )
-  drawQRCode(dynamicQrCode, qrCodeXPdf, qrCodeYPdf, qrCodeSizePdf, page)
+  drawQRCode(qrCodeContent, qrCodeXPdf, qrCodeYPdf, qrCodeSizePdf, page, false)
 }
 
-export async function generatePdf(activationCodes: DynamicActivationCode[], region: Region, pdfConfig: PdfConfig) {
+export async function generatePdf(
+  activationCodes: DynamicActivationCode[],
+  staticCodes: StaticVerificationCode[] | null,
```
Comment (michael-markl) [inline administration/src/cards/PdfFactory.ts:112]: Can we change the signature to something like `cards: { dynamicActivationCode: DynamicActivationCode, staticVerificationCode | null }[]` ?

If not, please add a check that the arrays have the same length

**Review by michael-markl — APPROVED** [review]

**Pull request merged**